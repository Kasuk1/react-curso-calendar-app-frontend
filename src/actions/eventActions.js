import Swal from 'sweetalert2';
import { fetchWithToken } from '../helpers/fetch';
import { prepareEvents } from '../helpers/prepareEvents';
import { types } from '../types/types';

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const resp = await fetchWithToken('events', event, 'POST');
      const json = await resp.json();
      if (json.ok) {
        event.id = json.event.id;
        event.user = {
          _id: uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
    } catch (err) {
      console.log(err);
    }
  };
};

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent,
});

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const json = await resp.json();

      if (json.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire('Error', json.msg, 'error');
      }
    } catch (err) {
      console.log(err);
    }
  };
};

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event,
});

export const eventStartDelete = (eventId) => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken(`events/${eventId}`, {}, 'DELETE');
      const json = await resp.json();
      if (json.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire('Error', json.msg, 'error');
      }
    } catch (err) {
      console.log(err);
    }
  };
};

const eventDeleted = () => ({
  type: types.eventDeleted,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchWithToken('events');
      const json = await resp.json();
      const events = prepareEvents(json.events);

      if (json.ok) {
        dispatch(eventLoaded(events));
      }
    } catch (err) {
      console.log(err);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});

export const eventLogout = () => ({
  type: types.eventLogout,
});
