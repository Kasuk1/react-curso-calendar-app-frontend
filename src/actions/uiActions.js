import { types } from '../types/types';

export const openModal = () => ({
  type: types.uiOpenModal,
});

export const closeModal = () => ({
  type: types.uiCloseModal,
});
