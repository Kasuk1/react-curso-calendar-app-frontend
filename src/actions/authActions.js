import Swal from 'sweetalert2';
import { fetchNoToken, fetchWithToken } from '../helpers/fetch';
import { types } from '../types/types';
import { eventLogout } from './eventActions';

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const resp = await fetchNoToken('auth', { email, password }, 'POST');
    const json = await resp.json();

    if (json.ok) {
      localStorage.setItem('token', json.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(
        login({
          uid: json.uid,
          name: json.name,
        })
      );
    } else {
      Swal.fire('Error', json.msg, 'error');
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const resp = await fetchNoToken(
      'auth/new',
      { name, email, password },
      'POST'
    );
    const json = await resp.json();

    if (json.ok) {
      localStorage.setItem('token', json.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(
        login({
          uid: json.uid,
          name: json.name,
        })
      );
    } else {
      Swal.fire('Error', json.msg, 'error');
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchWithToken('auth/renew');
    const json = await resp.json();

    if (json.ok) {
      localStorage.setItem('token', json.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(
        login({
          uid: json.uid,
          name: json.name,
        })
      );
    } else {
      dispatch(checkingFinish());
    }
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const checkingFinish = () => ({
  type: types.authCheckingFinish,
});

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(logout());
    dispatch(eventLogout());
  };
};

const logout = () => ({
  type: types.authLogout,
});
