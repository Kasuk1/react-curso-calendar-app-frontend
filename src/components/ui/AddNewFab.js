import React from 'react';
import { useDispatch } from 'react-redux';
import { openModal } from '../../actions/uiActions';

export const AddNewFab = () => {
  const dispatch = useDispatch();

  const handleAddClick = () => {
    dispatch(openModal());
  };

  return (
    <button className='btn btn-primary fab' onClick={handleAddClick}>
      <i className='fas fa-plus'></i>
    </button>
  );
};
