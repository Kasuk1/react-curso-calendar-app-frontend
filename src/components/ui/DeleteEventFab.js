import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { eventStartDelete } from '../../actions/eventActions';

export const DeleteEventFab = () => {
  const dispatch = useDispatch();
  const {
    activeEvent: { id: activeEventId },
  } = useSelector((state) => state.calendar);

  const handleDelete = () => {
    dispatch(eventStartDelete(activeEventId));
  };

  return (
    <button className='btn btn-danger fab-danger' onClick={handleDelete}>
      <i className='fas fa-trash'></i>
      <span> Borrar evento</span>
    </button>
  );
};
