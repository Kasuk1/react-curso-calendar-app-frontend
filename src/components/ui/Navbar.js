import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { startLogout } from '../../actions/authActions';

export const Navbar = () => {
  const dispatch = useDispatch();
  const { name } = useSelector((state) => state.auth);

  const handleLogout = () => {
    dispatch(startLogout());
  };

  return (
    <div className='navbar navbar-dark bg-dark mb-4'>
      <span className='navbar-brand'>{name}</span>
      <button className='btn btn-outline-danger'>
        <i className='fa fa-sign-out-alt'></i>
        <span onClick={handleLogout}> Salir</span>
      </button>
    </div>
  );
};
