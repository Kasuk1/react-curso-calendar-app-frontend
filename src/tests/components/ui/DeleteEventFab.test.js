import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';

import consfigureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { DeleteEventFab } from '../../../components/ui/DeleteEventFab';
import { eventStartDelete } from '../../../actions/eventActions';

jest.mock('../../../actions/eventActions', () => ({
  eventStartDelete: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = consfigureStore(middlewares);

const initState = {
  calendar: {
    activeEvent: {
      id: '6249c9c933c17b7da5c7c41a',
    },
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <DeleteEventFab />
  </Provider>
);

describe('<DeleteEventFab /> Tests', () => {
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call eventStartDelete on click', () => {
    wrapper.find('button').prop('onClick')();

    expect(eventStartDelete).toHaveBeenCalled();
  });
});
