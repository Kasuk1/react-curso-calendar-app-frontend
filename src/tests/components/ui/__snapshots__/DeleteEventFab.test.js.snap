// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeleteEventFab /> Tests should render correctly 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fas fa-trash"
  />
  <span>
     Borrar evento
  </span>
</button>
`;
